Feat/todo list

majishanred opened the pull request:
Тудушка готова

majishanred committed:
deps: добавил styled components

majishanred committed:
feat(Todo List): Сделал разметку листа, добавил
функционал и стили через Styled Components

myaak commented on package.json:
```diff
@@ -10,15 +10,19 @@
     "preview": "vite preview"
   },
   "dependencies": {
+    "@types/uuid": "^9.0.8",
```
Почему типы вообще в этих депсах. Эти депсы идут в сборку, в которой у тебя в итоге не будет ТСа. Такие вещи нужно пихать в devDeps. т.к они используются только в разработке

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,109 @@
+  isDone: boolean;
+};
+
+const StyledList = styled.ul`
```
Такие вещи можно выносить в отдельные файлы для чистоты. Или писать их в самом низу под компонентом Сверху они засоряют много места. Обычно делаем файлик `Component.styled`

IvanKoigerov replied:
ещё лучше в компонентах уносить эти вещи вниз компоненты, а не писать сверху
Также пишем `styled('ul')`, для общей идентичности

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,109 @@
+
+  const [filterInput, setFilterInput] = useState('');
+
+  const filteredTasks = filterInput
```
По хорошему такие вещи в useMemo закидывать.

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,109 @@
+    ? tasks.filter((elem) => elem.task.toLowerCase().includes(filterInput.toLowerCase()))
+    : tasks;
+
+  function onAdd(todo_item: TodoListElemBody) {
```
Т.к ты передаешь это вниз у тебя реденрятся деточки. Нужно исправить. Ну и по хорошему все-таки стрелками привыкать пользоваться. Чтобы хоистинга не было лишнего

majishanred replied:
Поменял на стрелки, но не очень понял 1 предложение, так что добавил useСallback для этих функций и memo для деток.

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,57 @@
+  display: flex;
+  align-items: center;
+
+  & > button {
```
А если потом я захочу другую кнопку импортанты писать?

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,57 @@
+          placeholder="Print here!"
+        ></input>
+      </label>
+      <button onClick={() => handleOnCreate()} type="button">
```
Тут можно и без `() =>` обойтись.
`onClick={handleCreate}`

myaak commented on src/components/TodoListElem.tsx:
```diff
@@ -0,0 +1,119 @@
+  onUpdate,
+  onDelete,
+}: {
+  todo_item: TodoElemBody;
```
Такие вещи лучше в отдельные типы выносить. Существуют компоненты с десятками пропов

myaak commented on src/components/TodoListElem.tsx:
```diff
@@ -0,0 +1,119 @@
+}) {
+  const [isEdit, setIsEdit] = useState<boolean>(false);
+
+  const handleOnSaveClick: FormEventHandler = (e) => {
```
Тут стрелку используешь. А почему в других местах fun declaration. Определись ( везде стрелка )

myaak commented on src/components/TodoListElem.tsx:
```diff
@@ -0,0 +1,119 @@
+    onUpdate(updatedTodo);
+  };
+
+  const handleOnEditClick = () => setIsEdit(true);
+  const handleOnCancleClick = () => setIsEdit(false);
+  const handleOnDeleteClick = () => onDelete(todo_item);
```
Какой смысл в этих функциях?

myaak commented on src/components/TodoListElem.tsx:
```diff
@@ -0,0 +1,119 @@
+  if (isEdit) {
+    return (
+      <TodoElemStyledForm method="post" onSubmit={handleOnSaveClick}>
+        <input name="taskInput" defaultValue={todo_item.task}></input>
```
Зачем делать тег двойным? `<input />`

myaak commented on src/components/TodoListElem.tsx:
```diff
@@ -0,0 +1,119 @@
+      task: formJson.taskInput as string,
+    };
+
+    setIsEdit(false);
+    onUpdate(updatedTodo);
```
А теперь представим что у тебя на апдейт асинхронное действие ( на бек отправляешь ). Получается я уже не в эдите и не знаю че там с моей тудухой?

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,109 @@
+    const newTodo = { ...todo_item, id: uuidv1() };
+    setTasks(() => [...tasks, newTodo]);
+  }
+
+  function onUpdate(todo_item: TodoListElem) {
```
Определись со стилем переменных.
Все пишем в camelCase. У тебя где-то camel, где-то snake_case.

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,109 @@
+
+  function onAdd(todo_item: TodoListElemBody) {
+    const newTodo = { ...todo_item, id: uuidv1() };
+    setTasks(() => [...tasks, newTodo]);
```
Очень плохо, а если нужно две тудушки добавить последовательно?

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,109 @@
+    const newArr = [...tasks];
+    newArr[index] = todo_item;
+
+    setTasks(() => [...newArr]);
```
Какой смысл от колбека и спреда тут?

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,57 @@
+    };
+
+    setTodoInput('');
+    onSave(newTodo);
```
Собственно тож самое. Инпут уже почистился, а запрос еще идет

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,109 @@
+import { v1 as uuidv1 } from 'uuid';
+import styled from 'styled-components';
+
+export type TodoListElem = {
```
Обычно такие штуки называют TodoItem или просто Todo.

Но чаще описываем интерфейсами и пишем `ITodo`

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,57 @@
+  };
+
+  return (
+    <StyledForm id="todoInsert">
```
А где id используется, не нашел

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,57 @@
+        <input
+          name="taskInput"
+          value={todoInput}
+          onChange={(e) => setTodoInput(e.target.value)}
```
Чтобы в таких случаях писать меньше приколов можно доставать `e.target.value` в компоненте и написать просто
`onChange={setTodoInput}`

myaak commented on src/components/TodoListElem.tsx:
```diff
@@ -0,0 +1,119 @@
+  display: flex;
+  flex-direction: row;
+
+  & > div {
```
Аааааа, плохо на селекторы вешать такие вещи.

myaak commented on src/components/TodoListElem.tsx:
```diff
@@ -0,0 +1,119 @@
+      <input
+        name="inputCheckbox"
+        type="checkbox"
+        checked={todo_item.isDone ? true : false}
```
Зачем. у тебя isDone и так boolean.
`checked={todo_item.isDone}`

myaak commented on src/components/TodoListElem.tsx:
```diff
@@ -0,0 +1,119 @@
+  const handleOnSaveClick: FormEventHandler = (e) => {
+    e.preventDefault();
+
+    const form = e.target as HTMLFormElement;
+    const formData = new FormData(form);
+    const formJson = Object.fromEntries(formData.entries());
```
Интересный подход, но можно было и стейтами сделать.

IvanKoigerov replied:
Да и лучше сделай на стейте, хотел бы именно такую реализацию увидеть

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,57 @@
+    background-color: hsl(180, 29%, 50%);
+    color: #ffffff;
+
+    padding: 10px 15px;
```
Странные цифры паддинга, даже не четные

IvanKoigerov commented:
Ещё кстати, тк ты побежал дальше нужного, ты не знаешь какой стек нужно использовать))
Одну библиотеку ты тут забыл) 
Так что нужно будет переписывать, а вот какую)))
Как ты думаешь?

IvanKoigerov commented on src/components/TodoListElem.tsx:
```diff
@@ -0,0 +1,119 @@
+  const handleOnSaveClick: FormEventHandler = (e) => {
+    e.preventDefault();
+
+    const form = e.target as HTMLFormElement;
```
Смысл тут типизировать, типизируй в `(e: Type) =>`, as используем прям в крайнем случае, это не красиво, не всегда читабельно, а также в этой ситуации ты можешь передать любую `e`, хоть стрингу, а это плохо

IvanKoigerov commented:
Выложи свой пр на https://stackblitz.com, чтобы потыкаться можно было

majishanred committed:
feat(Todo List): Изменил разметку компонентов,
раскидал Styled Components в отдельную директорию,
перекинул тип в отдельную директорию

majishanred committed:
feat(Todo List): Перекинул uuid в дев зависимости,
выделил пропс из TodoItem

majishanred commented:
https://stackblitz.com/~/github.com/majishanred/webbee-todo

majishanred committed:
refactor(TodoList):
Добавил useCallback для функций в App.tsx,
Применил memo для рендера TodoListItem и
TodoInputForm

majishanred committed:
refactor(Todo Item Interface):
Поменял type на interface

majishanred committed:
refactor(TodoList): Добавил к onAdd возможность
обрабатывать массив элементов

majishanred committed:
Refactor(TodoInputForm): поменял onChange handler

majishanred commented:
> Ещё кстати, тк ты побежал дальше нужного, ты не знаешь какой стек нужно использовать)) Одну библиотеку ты тут забыл) Так что нужно будет переписывать, а вот какую))) Как ты думаешь?

MUI?

majishanred commented:
@IvanKoigerov @myaak

myaak commented:
Да, должен быть MUI @majishanred

myaak commented on src/interfaces/ITodoItem.ts:
```diff
@@ -0,0 +1,7 @@
+  isDone: boolean;
+}
+
+export interface ITodoItemBody extends Omit<ITodoItem, 'id'> {}
```
Такие вещи обычно называются EntityCreate для этого примера `ITodoItemCreate` или просто `TodoItemCreate`

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,80 @@
+
+  const [filterInput, setFilterInput] = useState('');
+
+  const filteredTasks = useMemo(
+    () => (filterInput ? tasks.filter((elem) => elem.task.toLowerCase().includes(filterInput.toLowerCase())) : tasks),
+    [tasks, filterInput],
+  );
```
Вынесем по логике

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,46 @@
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const setInput = () => {
+    setTodoInput(inputRef.current!.value);
```
! Плохой тон. Сделай проверку. Он не просто так жалуется.
Например 
`if (!inputRef.current) {
     return;
}
set()
`

myaak reviewed (changes requested):
Давай сделаем три компонента.
1. Фильтр
2. ТудуЛист
3. Форма добавления туду

Очевидно ну можно разбить туду лист на todoItem, но должно быть три главных

majishanred committed:
deps: Убрал styled-components, добавил Mui, emotion and emotion/react

majishanred committed:
pr(TodoList): Добавил Mui и emotion,
разбил на 3 компонента, добавил debounce на change
фильтра

myaak commented on src/components/TodoFilter.tsx:
```diff
@@ -0,0 +1,16 @@
+  onFilterSet: Dispatch<string>;
+};
+
+export default function TodoFilter({ filter, onFilterSet }: TodoFilterProps) {
+  const handleOnChange = debounce((e: ChangeEvent<HTMLInputElement>) => {
+    onFilterSet(e.target.value);
+  }, 500);
```
А давай вынесем фильтр в контекст, чтобы при кнопке add не рендерился фильтр.
Фильтр должен рендериться только тогда когда ты пишешь в него.
А так он рендерится когда добавляешь таску тоже.

Давай вынесем фильтр в контекст.

myaak commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,83 @@
+import { ITodoItem } from '../interfaces/ITodoItem';
+import { Box, BoxProps, Button, ButtonGroup, Checkbox, TextField, Typography, styled } from '@mui/material';
+
+type TodoElementProps = {
+  todoItem: ITodoItem;
+  onUpdate: Dispatch<ITodoItem>;
+  onDelete: Dispatch<ITodoItem>;
+};
```
Раз есть отдельный файл с типами, то давай туда.

myaak commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,83 @@
+
+  return (
+    <TodoStyledBox component="div">
+      <Checkbox checked={todoItem.isDone} onChange={handleCheckboxClick} />
```
Можно todoItem деструктурировать в начале, чтобы не писать постоянно

myaak commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  return (
+    <TodoStyledBox component="div">
```
Box и есть div

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,34 @@
+  };
+
+  const handleOnCreate = () => {
+    const newTodo = {
+      task: input,
+      isDone: false,
+    };
```
Можно написать какую-то константу или тут имплисит задать, чтоб если у тебя в типах поле поменяется ТС ругался

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,66 @@
+  const [filterInput, setFilterInput] = useState('');
+
+  const onAdd = useCallback((newTodo: ITodoItemCreate | ITodoItemCreate[]) => {
+    if (newTodo instanceof Array) {
+      const newTodos = newTodo
+        .filter((elem) => elem.task)
+        .map((elem) => {
+          return { ...elem, id: uuidv1() };
+        });
+
+      setTasks((tasks) => [...tasks, ...newTodos]);
+    } else {
+      if (!newTodo.task) return;
+      const elemToAdd = { ...newTodo, id: v1() };
+      setTasks((tasks) => [...tasks, elemToAdd]);
+    }
```
У тебя вроде не создается несколько тасок. Зачем проверка на массив? К тому же можно провеить Array.isArray(newTodo)

majishanred replied:
Ты мне замечаниях оставил:
![image](https://github.com/majishanred/webbee-todo/assets/58140063/8114348e-eb4c-4532-9632-edd207cfd202)

myaak replied:
оставил замечание что ты не колбеком из предыдущего значения пушишь, а используешь текущее значение стейта.

myaak commented on src/App.tsx:
```diff
@@ -1,7 +1,66 @@
+
+      setTasks((tasks) => [...tasks, ...newTodos]);
+    } else {
+      if (!newTodo.task) return;
```
Такие проверки лучше вообще в самый верх выносить, чтобы вообще не идти дальше по функции

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,34 @@
+  const [input, setInput] = useState('');
+
+  const handleOnChange = (e: ChangeEvent<HTMLInputElement>) => {
+    setInput(e.target.value);
```
Вообще можно и так. А можно прям внутри textField написать. Не криминал

myaak commented on src/components/TodoList.tsx:
```diff
@@ -0,0 +1,25 @@
+  return (
+    <>
+      {filteredTasks.map((todo) => (
+        <TodoListElem key={todo.id} todoItem={todo} onUpdate={onUpdate} onDelete={onDelete}></TodoListElem>
```
Пройдись везде по компонентам. Тут вот например нет смысла в двойном теге. Пиши `<TodoListElem />`, а еще нейминг должен быть наверное TodoItem), а не просто файл

majishanred committed:
pr: поправил согласно замечаниям

myaak commented on src/components/TodoFilter.tsx:
```diff
@@ -0,0 +1,16 @@
+import debounce from 'lodash.debounce';
+import UpdateFilterContext from '../contexts/updateFilterContext';
+
+const TodoFilter = memo(function TodoFilter() {
```
какое мемо

majishanred committed:
pr: Поправил в соответствии с замечанием

majishanred committed:
pr: Поправил нейминг

majishanred committed:
pr: Убрал ненужное .memo(), добавил 2 конеткста и провайдер для них

majishanred committed:
pr: поправил нейминг контекстов

majishanred committed:
pr: поправил контекст Todo, чтобы было четкое разделение на read и write

majishanred committed:
pr: добавил reducer для управления TodoList

majishanred committed:
pr: убрал console.log

majishanred committed:
pr: убрал ненужную ошибку

myaak commented on src/components/TodoList.tsx:
```diff
@@ -0,0 +1,24 @@
+  return (
+    <>
+      {filteredTasks.map((todo) => (
+        <TodoItem key={todo.id} todoItem={todo} onUpdate={dispatch} onDelete={dispatch} />
```
Зачем передаешь пропом? У тебя это в контексте есть

myaak commented on src/components/TodoFilter.tsx:
```diff
@@ -0,0 +1,16 @@
+import { TextField } from '@mui/material';
+import { ChangeEvent, useContext } from 'react';
+import debounce from 'lodash.debounce';
+import UpdateFilterContext from '../contexts/UpdateFilterContext';
```
Ошибки при запуске проекта, у тебя тут с большой буквы, а файл с маленькой.

myaak commented on src/contexts/ReadTodoContext.ts:
```diff
@@ -0,0 +1,4 @@
+import { createContext } from 'react';
+import { ITodoItem } from '../interfaces/ITodoItem';
+
+export const TodoReadContext = createContext<ITodoItem[]>([]);
```
Не совсем вижу смысл в такой структуре. Я бы прям в файле объявил оба контекста и там же написал провайдер для них.

myaak commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,37 @@
+      isDone: false,
+    };
+
+    dispatch({ todo: newTodo, type: 'add' });
```
у тебя этот dispatch по факту setState. Не понятно к чему он относится. Напиши хотя бы `dispatchTodoList`, или вообще setTodoList. Да и type наверное нет смысла юзать.
Зачем мне писать
`dispatch({ todo: newTodo, type: 'add' }); dispatch({ todo: newTodo2, type: 'add' });`, если хочу добавить другую тудушку еще (вдруг).

myaak commented on src/contexts/WriteTodoContext.ts:
```diff
@@ -0,0 +1,4 @@
+import { Dispatch, createContext } from 'react';
+import { TodoAction } from '../reducers/TodoReducer';
+
+export const WriteTodoContext = createContext<Dispatch<TodoAction>>(() => {});
```
Обычно ему задают дефолт значение `null` или `{}`

myaak commented on src/prodviders/TodoProvider.tsx:
```diff
@@ -0,0 +1,29 @@
+];
+
+export const TodoProvider = ({ children }: PropsWithChildren) => {
+  const [tasks, dispatchActions] = useReducer(TodoReducer, defaultTasks);
```
useReducer не думаю что есть лютый смысл использовать. Думаю тут useState нормально бы справился. Го переделаем

majishanred committed:
pr: убрал ненужные провайдеры, контексты и редуктор

majishanred committed:
pr: поправил контексты и провайдеры, убрал редуктор, переделал туду контекст на использование стейта

myaak commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,102 @@
+    return (
+      <TodoStyledBox>
+        <TextField value={input} onChange={handleChange} />
+        <ButtonGroup disableRipple={true} sx={{ marginLeft: 'auto' }}>
```
true по дефолту.
можно написать просто `disableRipple`

myaak commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,102 @@
+    });
+  };
+
+  const handleOnSaveClick = () => {
+    const updatedTodo: ITodoItem = {
+      ...todoItem,
+      task: input,
+    };
+
+    onUpdate(updatedTodo);
+    setIsEdit(false);
+  };
```
Наверное по хорошему это в одно место вынести, а то у тебя делит в одном, создание в другом компоненте. (ИМХО)

IvanKoigerov commented on src/components/TodoFilter.tsx:
```diff
@@ -0,0 +1,17 @@
+  const setFilter = useContext(WriteFilterContext);
+
+  const handleOnChange = debounce((e: ChangeEvent<HTMLInputElement>) => {
+    if (!setFilter) return;
```
Лучше создавать обертку, которая будет ошибки выкидывать, чтобы не делать такие проверки

IvanKoigerov commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,40 @@
+import { ChangeEvent, useContext, useState } from 'react';
+import { Box, Button, ButtonGroup, TextField } from '@mui/material';
+import { WriteTodoContext } from '../contexts/TodoContext';
+import { v1 as uuidv1 } from 'uuid';
```
Я бы на твоем месте не юзал бы библу для уникальных id
А просто сам создавал id

IvanKoigerov commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,40 @@
+const TodoInputForm = () => {
+  const [input, setInput] = useState('');
+
+  const saveTodos = useContext(WriteTodoContext);
```
save, это сохранение, лучше set или create

IvanKoigerov commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,40 @@
+    setInput(e.target.value);
+  };
+
+  const handleOnCreate = () => {
```
будто бы нет смысла добавлять слово `handle`, обычно редко так делаем, но это нормально

IvanKoigerov commented on src/components/TodoInputForm.tsx:
```diff
@@ -0,0 +1,40 @@
+
+  return (
+    <Box display="flex" alignItems="center">
+      <TextField placeholder="Type task here" value={input} onChange={handleOnChange} sx={{ width: '70%' }} />
```
Не очень круто использовать % значение ширены и высоты, 100% и 50% можем спокойно использовать. А вот такие вещи уже не очень, не совсем понятно каким он размером по итогу будет

IvanKoigerov commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,102 @@
+import { TodoItemProps } from '../interfaces/TodoItemProps';
+import { WriteTodoContext } from '../contexts/TodoContext';
+
+const TodoItem = ({ todoItem }: TodoItemProps) => {
+  const { task, isDone } = todoItem;
```
А почему просто не передавать {task, isDone} в пропы, чтобы вот так не распаковывать

myaak replied:
Если пропов станет 100 - будет лень. Лучше передать фул, как-будто

IvanKoigerov commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,102 @@
+  const saveTodos = useContext(WriteTodoContext);
+
+  const onDelete = () => {
+    if (!saveTodos) return;
+    saveTodos((todos) => todos.filter((elem) => elem.id !== todoItem.id));
+  };
+
+  const onUpdate = (updatedTodo: ITodoItem) => {
+    if (!saveTodos) return;
+
```
Ну и тут также, можно сделать обертку, которая будет выкидывать ошибку если нет значения

IvanKoigerov commented on src/components/TodoList.tsx:
```diff
@@ -0,0 +1,22 @@
+  const filter = useContext(ReadFilterContext);
+  const todoList = useContext(ReadTodoContext);
+
+  const filteredTasks = useMemo(
+    () => (filter ? todoList.filter((elem) => elem.task.toLowerCase().includes(filter.toLowerCase())) : todoList),
+    [todoList, filter],
+  );
```
лучше такие вещи расписывать нормально с if, иначе теряется читабельность

IvanKoigerov commented on src/interfaces/ITodoItem.ts:
```diff
@@ -0,0 +1,5 @@
+export interface ITodoItem {
+  id: string;
+  task: string;
+  isDone: boolean;
+}
```
`I` устарела, обычно так не описывают сейчас
Сейчас используют либо `Props`, если прокидываешь в компонент, а если как тип какого-то значение, то просто название сущности, ну и если нужно можно добавить `Type`

В твоем варианте самое корректное название это просто `ToDo`

IvanKoigerov replied:
Сейчас не парься, не меняй, просто на будущее.

IvanKoigerov commented:
Пару моментов поправь, напиши мне и потом можешь начинать делать todo c state Manager в новой ветку, от этой

majishanred committed:
pr(todo): поправил изменения

IvanKoigerov commented on src/utils.ts:
Круто
